datumaro/plugins/synthetic_data/utils.py

```diff
@@ -0,0 +1,198 @@
+                    break
+
+            self.xs.append(next_x)
+            self.ys.append(next_y)
+            prev_x = next_x
+            prev_y = next_y
+
+    @staticmethod
+    def process_nans(data):
+        nan_index = np.nonzero(np.isnan(data))
+        extend = np.array(range(nan_index[0][0] - 100, nan_index[0][0]))
+        delete_row = np.append(extend, nan_index)
+        return delete_row
+
+    def rescale(self, image_x, image_y, pad_x, pad_y):
+        if image_x < 2 * pad_x or image_y < 2 * pad_y:
+            raise ValueError(
+                f"Image generation with height < {2 * pad_x} or width < {2 * pad_y} is not supported"
+            )
+
+        xs = np.array(self.xs)
+        ys = np.array(self.ys)
+        if np.any(np.isnan(xs)):
+            delete_row = self.process_nans(xs)
+            xs = np.delete(xs, delete_row, axis=0)
+            ys = np.delete(ys, delete_row, axis=0)
+
+        if np.any(np.isnan(ys)):
+            delete_row = self.process_nans(ys)
+            xs = np.delete(xs, delete_row, axis=0)
+            ys = np.delete(ys, delete_row, axis=0)
+
+        if np.min(xs) < 0.0:
+            xs -= np.min(xs)
+        if np.min(ys) < 0.0:
+            ys -= np.min(ys)
+        xmax, xmin = np.max(xs), np.min(xs)
+        ymax, ymin = np.max(ys), np.min(ys)
+        self.xs = np.uint16(xs / (xmax - xmin + 1e-5) * (image_x - 2 * pad_x) + pad_x)
+        self.ys = np.uint16(ys / (ymax - ymin + 1e-5) * (image_y - 2 * pad_y) + pad_y)
+
+    def draw(self, image_x, image_y, draw_point, pad_x=6, pad_y=6):
+        self.rescale(image_x, image_y, pad_x, pad_y)
+        image = np.zeros((image_x, image_y), dtype=np.uint8)
+        for x, y in zip(self.xs, self.ys):
+            if draw_point:
+                image[x, y] = 127
+            else:
+                mask = "{:09b}".format(self._rng.randint(1, 511))
+                patch = 127 * np.array(list(map(int, list(mask))), dtype=np.uint8).reshape(3, 3)
+                image[x + 1 : x + 4, y + 1 : y + 4] = patch
+
+        return image
+
+
+def download_colorization_model(path):
+    proto_name = "colorization_deploy_v2.prototxt"
+    model_name = "colorization_release_v2.caffemodel"
+    npy_name = "pts_in_hull.npy"
+
+    if not (
+        osp.exists(osp.join(path, proto_name))
+        and osp.exists(osp.join(path, model_name))
+        and osp.exists(osp.join(path, npy_name))
+    ) and not os.access(path, os.W_OK):
+        raise ValueError(
+            "Please provide path where colorization model is located or path to writable directory to save colorization model"
+        )
+
+    if not osp.exists(osp.join(path, proto_name)):
+        log.warning(
+            "Downloading '%s' model for image coloring to '%s'",
+            proto_name,
+            path,
+        )
+        url = "https://raw.githubusercontent.com/richzhang/colorization/caffe/colorization/models/"
+        proto = requests.get(url + proto_name)
```

Comment:
There are several problems here:

1. The naming of the variable is questionable, because this isn't the proto file - it's a response object. I would suggest renaming it to `response`.
2. You must check whether the request actually succeeded. You can do that with `response.raise_for_status()`.
3. You need some form of integrity checking. I would suggest hardcoding a SHA-512 digest for each file and comparing the digest against the downloaded file before saving it.
4. Using `requests.get` is inadvisable, because then Requests won't know when to close the session. Use `requests.Session` as a context manager and use `session.get` instead.

Reply:
Updated, +1 for content verification.

Reply:
Fixed

Reply:
You didn't fix item 4.

Reply:
I didn't find any confirmation to this in the library docs. Could you provide a link and the situation, when it's a problem?

Reply:
Huh. It turns out that this issue I was talking about was actually fixed years ago, in psf/requests@3155bc99362a8c6ab136b6a3bb999732617cd2e5. Good to know.

That said, using a session would still be beneficial, because two of the files are downloaded from the same host, and a session would allow the connection to be reused.